ethcore/engines/validator-set/src/contract.rs

```diff
@@ -48,31 +49,64 @@ pub struct ValidatorContract {
 	contract_address: Address,
 	validators: ValidatorSafeContract,
 	client: RwLock<Option<Weak<dyn EngineClient>>>, // TODO [keorn]: remove
+	posdao_transition: Option<BlockNumber>,
 }
 
 impl ValidatorContract {
-	pub fn new(contract_address: Address) -> Self {
+	pub fn new(contract_address: Address, posdao_transition: Option<BlockNumber>) -> Self {
 		ValidatorContract {
 			contract_address,
-			validators: ValidatorSafeContract::new(contract_address),
+			validators: ValidatorSafeContract::new(contract_address, posdao_transition),
 			client: RwLock::new(None),
+			posdao_transition,
 		}
 	}
 }
 
 impl ValidatorContract {
-	fn transact(&self, data: Bytes) -> Result<(), String> {
-		let client = self.client.read().as_ref()
-			.and_then(Weak::upgrade)
-			.ok_or_else(|| "No client!")?;
-
-		match client.as_full_client() {
-			Some(c) => {
-				c.transact(TransactionRequest::call(self.contract_address, data))
-					.map_err(|e| format!("Transaction import error: {}", e))?;
-				Ok(())
-			},
-			None => Err("No full client!".into()),
+	fn transact(&self, data: Bytes, gas_price: Option<U256>) -> Result<(), String> {
+		let client = self.client.read().as_ref().and_then(Weak::upgrade).ok_or("No client!")?;
+		let full_client = client.as_full_client().ok_or("No full client!")?;
+		let tx_request = TransactionRequest::call(self.contract_address, data).opt_gas_price(gas_price);
+		match full_client.transact(tx_request) {
+			Ok(()) | Err(transaction::Error::AlreadyImported) => Ok(()),
+			Err(e) => Err(e.to_string())?,
+		}
+	}
+
+	fn do_report_malicious(&self, address: &Address, block: BlockNumber, proof: Bytes) -> Result<(), EthcoreError> {
```

Comment:
Maybe?
```suggestion
	fn enqueue_malicious_behaviour_report(&self, address: &Address, block: BlockNumber, proof: Bytes) -> Result<(), EthcoreError> {
```

Reply:
But we not only enqueue, but also send a transaction:
```rust
self.transact(data, gas_price)?;
```
The queue is only a fallback, so that we retry if the transaction fails to get included in a block.